src/cmap/wire_protocol/on_demand/document.ts

```diff
@@ -0,0 +1,315 @@
+import {
+  Binary,
+  BSON,
+  type BSONElement,
+  BSONError,
+  type BSONSerializeOptions,
+  BSONType,
+  getBigInt64LE,
+  getFloat64LE,
+  getInt32LE,
+  ObjectId,
+  parseToElementsToArray,
+  Timestamp,
+  toUTF8
+} from '../../../bson';
+
+// eslint-disable-next-line no-restricted-syntax
+const enum BSONElementOffset {
+  type = 0,
+  nameOffset = 1,
+  nameLength = 2,
+  offset = 3,
+  length = 4
+}
+
+export type JSTypeOf = {
+  [BSONType.null]: null;
+  [BSONType.undefined]: null;
+  [BSONType.double]: number;
+  [BSONType.int]: number;
+  [BSONType.long]: bigint;
+  [BSONType.timestamp]: Timestamp;
+  [BSONType.binData]: Binary;
+  [BSONType.bool]: boolean;
+  [BSONType.objectId]: ObjectId;
+  [BSONType.string]: string;
+  [BSONType.date]: Date;
+  [BSONType.object]: OnDemandDocument;
+  [BSONType.array]: OnDemandDocument;
+};
+
+/** @internal */
+export class OnDemandDocument {
+  /** Caches the existence of a property */
+  private readonly existenceOf: Record<string, boolean> = Object.create(null);
+  /** Caches a look up of name to element */
+  private readonly elementOf: Record<string, BSONElement> = Object.create(null);
+  /** Caches the revived javascript value */
+  private readonly valueOf: Record<string, any> = Object.create(null);
```

Comment:
Since a primary goal of `OnDemandDocument` is to provide a more performant interface for accessing BSON, it might be worth grouping these into a single `Record<string, { element: BSONElement; value: any; ... }>` to avoid creating multiple dictionary objects, because currently accessing a single subdocument already creates at least 6 new objects instead of 1 :)

Reply:
Also related to this and https://github.com/mongodb/node-mongodb-native/pull/4061#discussion_r1544748906 - could we simply store whether or not we've deserialized the whole BSON?  That would remove the need for `existenceOf` entirely.

Reply:
I think I can manage to collapse the existence, element, and value cache's into one. I think the indexFound one will need to remain seperate. I went in this direction because implementing the cache that way creates a new object per element as it is found, but I guess an allocation is an allocation, so it doesn't matter much might as well consolidate. I'll give that a try

> could we simply store whether or not we've deserialized the whole BSON?

Can you expand on what you invision here? (specifcally "whole BSON") do you mean caching the result of a toObject call or tracking that we've finished finding all elements?

Reply:
I removed existence, value, and element cache, in favor of one that tracks all of those pieces of work. The `cache[name]` being `false` is not checked outside of the getElement function so as long as any methods go through getElement the same nullish checks are used for the element itself and its potentially revived value.